src/collapse/perf.rs

```diff
@@ -324,6 +354,87 @@ impl PerfState {
         }
         Ok(())
     }
+
+    // Use addr2line to determine the symbol to use for each program counter address
+    // (as opposed to using the symbol names that perf script produces)
+    fn un_inline(&mut self, addr: &str, module: &str) -> Result<(), String> {
+        let addr = u64::from_str_radix(addr, 16)
+            .map_err(|_| format!("unable to parse {} as hex address", addr))?;
+
+        if let Entry::Vacant(entry) = self.addr2line_contexts.entry(module.into()) {
+            let file =
+                File::open(module).map_err(|_| format!("unable to open file: {}", module))?;
+            // Using memmap is unsafe because the underlying file may be modified,
+            // which would cause undefined behavior. It's unlikely to be a problem in practice
+            // since we shouldn't run very long. The potential speed-up is worth it.
+            let map = unsafe {
+                memmap::Mmap::map(&file)
+                    .map_err(|_| format!("unable to memmap file: {}", module))?
+            };
+            let file = &object::File::parse(&*map)
+                .map_err(|_| format!("unable to parse file: {}", module))?;
+            entry.insert(addr2line::Context::new(file).map_err(|e| e.to_string())?);
+        };
+        let ctx = &self.addr2line_contexts[module];
+
+        let mut frames = ctx.find_frames(addr).map_err(|_| {
+            format!(
+                "unable to parse frames from module: {}, at address: {}",
+                module, addr
+            )
+        })?;
+
+        let mut funcs = SmallVec::<[String; 1]>::new();
+        while let Some(frame) = frames.next().unwrap() {
+            let func = frame
+                .function
+                .map(|func| {
+                    let name = func.raw_name().map(Cow::from);
+                    let name = name.unwrap_or_else(|_| Cow::from("??"));
+                    let name = remove_discriminator(name);
+                    match func.language {
+                        Some(gimli::DW_LANG_Rust) => {
+                            // Using rustc_demangle directly here since addr2line::demangle_auto
+                            // doesn't remove trailing hash value from Rust symbol names.
+                            // See https://github.com/gimli-rs/addr2line/issues/108
+                            format!("{:#}", rustc_demangle::demangle(&name))
+                        }
+                        lang => addr2line::demangle_auto(name, lang).to_string(),
+                    }
+                })
+                .unwrap_or_else(|| "??".into());
+
+            if self.opt.show_context {
+                let loc = frame.location;
+                let (file, line) = match loc {
```

Comment:
Can we just `match` directly on `frame.location` here?

Reply:
Looks like we can now. With the way I was getting the file name before, the compiler complained, but it's no longer an issue.